### ISSUE TITLE
Join forces on algorithms and manifolds?

### ISSUE THREAD

**Issue by kellertuer** [JuliaGNI/GeometricOptimizers.jl#4]
hi,

this is a post to reach out concerning the manifolds and algorithms used here. You do nt have a discussion page, I would have opened it there instead of as an issue.

We have the packages [Manopt.jl](https://manoptjl.org), which (as far as I see) covers all but the Adam() case of your solvers – e.g. 
* https://manoptjl.org/stable/solvers/gradient_descent/ for example also https://manoptjl.org/stable/solvers/gradient_descent/#Manopt.MomentumGradient but also a large collection of CGs https://manoptjl.org/stable/solvers/conjugate_gradient_descent/
* several variants of https://manoptjl.org/stable/solvers/quasi_Newton/
* ...

While [Manifolds.jl](https://github.com/JuliaManifolds/Manifolds.jl)
* Has https://juliamanifolds.github.io/Manifolds.jl/stable/manifolds/stiefel/ even with different metric
* Has https://juliamanifolds.github.io/Manifolds.jl/stable/manifolds/grassmann/ even in both Stiefel and projector representation.

Just that here both are in one repository, e.g. like in the Manopt (Matlab) variant, which is also following this more mono-repo style.

Would it maybe make sense to try to do a “manifold-generic” Adam variant based on your paper? 
And in general we could explore whether we can avoid reimplementing manifolds and instead focussing on on getting one implementation fast?

We for example have
* https://github.com/JuliaManifolds/LieGroups.jl

but are still vey early on the GPU side – just a first few steps in https://github.com/JuliaManifolds/ManifoldsGPU.jl

Of course I understand if you prefer to continue on your own, I just wanted to carefully reach out

## Status

This is exploratory package-collaboration outreach rather than an actionable repository task. Please continue this conversation in GitHub Discussions or a dedicated collaboration proposal if there is a concrete integration plan; otherwise this issue can remain closed.

**Comment by benedict-96** [comment]
Hello @kellertuer,

Thanks a lot for reaching out! `Manopt.jl` and `Manifolds.jl` look like very interesting packages, but I unfortunately did not have a chance to go through them in detail yet. For us `GeometricOptimizers.jl`'s primary use is training neural networks from [`GeometricMachineLearning.jl`](https://github.com/JuliaGNI/GeometricMachineLearning.jl), where the GPU compatibility is crucial. As you may have seen `GeometricOptimizers.jl` is very  much work in progress at the moment though.
We will look through your packages to see what the most straightforward approach would be going forward, but implementing Adam for various manifolds in `Manopt.jl` (the paper gives a general recipe for doing so for homogeneous spaces) may be a good option. 
Please note that at the moment we have Easter holidays, so a definite answer will probably take at least 2+ weeks.

**Comment by kellertuer** [comment]
Hi,

I did not check the package here too detailed, but the BFGS for example seems a bit of in early stage for example, but sure, as you write, this is early stage. But since it is early stage, I wanted to reach out now as well, since later it might be a huge rework if you would like to depend on either Manifolds.jl or Manopt.jl.

I did not yet have the time to check your paper on ADAM, the others that I did see (quite a while back when I implemented e.g. Nesterov and Momentum) none was generic enough for me, because I do like implementing stuff generically for most/all manifolds and not just for each individually, we have too many manifolds by now for that.

And no worries, I am travelling in Tokyo for the next 2 weeks, and if I would be home in Norway, there the easter week is a super important vacation ... also none of this is urgent, nor do we __have__ to join; if your goals are different and you feel, the features/Manifolds/Algorithms offered are too different in __whatever__ – we can also not do that.
I just feel that “reimplementing the wheel” (which I see a lot a lot in ML, but I am also not that much in ML myself) is not always necessary, when the old wheel is good enough.

But First: God påske | Frohe Ostern!

**Comment by mateuszbaran** [comment]
Hi! I'd like to add that GPU compatibility of Manopt itself is fairly good and has been verified for a number of solvers; see here for recent work: https://github.com/JuliaManifolds/Manopt.jl/pull/577 . There may still be a few places that might require patching, especially in more complicated solvers, though Manopt's architecture is generally GPU-friendly.

A more problematic part is GPU support for specific operations on specific manifolds. That part is largely being worked on in ManifoldsGPU.jl. I think we've managed to find a good structure for how to add GPU support but it's still far from completeness. At least it's being actively worked on.

**Comment by benedict-96** [comment]
Thanks for pointing that out @kellertuer. Yes, we definitely do not want to exactly rewrite code that already exists multiple times, a problem which many areas of scientific computing suffer from. We are usually also aiming for greater generality when we implement things, and how generic the interface has to be typically only becomes apparent after a few objects have been implemented.
Thanks for the comment @mateuszbaran! For GPU support we rely on `KernelAbstractions.jl`, but plan to also look into `JACC.jl`. By looking at the [Project file of `ManifoldsGPU.jl`](https://github.com/JuliaManifolds/ManifoldsGPU.jl/blob/main/Project.toml) it seems that you are first focusing on CUDA support?

Anyway, I'll take a closer look at your packages and we'll get back to you to let you know how we plan to proceed!

God påske, Frohe Ostern, Wesołych Świąt!

**Comment by mateuszbaran** [comment]
> Thanks for the comment [@mateuszbaran](https://github.com/mateuszbaran)! For GPU support we rely on `KernelAbstractions.jl`, but plan to also look into `JACC.jl`. By looking at the [Project file of `ManifoldsGPU.jl`](https://github.com/JuliaManifolds/ManifoldsGPU.jl/blob/main/Project.toml) it seems that you are first focusing on CUDA support?

Yes, we're currently focusing on CUDA but we'd definitely welcome other accelerators as well, as submodules of ManifoldsGPU. The biggest obstacle we've encountered so far is GPU support for matrix factorization, which is somewhat limited even in CUDA. Basically everything else can be approximated decently by array broadcasting, scalar special functions and matrix multiplication kernels.

Feel free to ask if you have any questions about JuliaManifolds 🙂 

Frohe Ostern!

**Comment by kellertuer** [comment]
> Yes, we definitely do not want to exactly rewrite code that already exists multiple times, a problem which many areas of scientific computing suffer from. We are usually also aiming for greater generality when we implement things, and how generic the interface has to be typically only becomes apparent after a few objects have been implemented.

That sounds great! And I do agree, one can not always reuse existing code – sometimes goals and approaches are too different; that is also why I tried to write the initial post as careful as possible :)

Concerning ManifoldsGPU.jl – we had that on our list for some time, but it sometimes needs a concrete reason to get started, which was someone coming by wanting to use CUDA. We basically use the same approach as you mentioned, but getting that to the generic case would also be very cool and welcome.

**Comment by michakraus** [comment]
My apologies. That was actually an accident (or rather a poor prompt that lead to an ambiguous job description).

Also apologies that we did not respond to your messages more properly more promptly. We briefly discussed the idea of joining forces after your original message, but couldn‘t come to a conclusion on how to best approach this. Life and other work kept both of us very busy lately, so that the question largely went under the radar.

Nonetheless, we are very much in favour of not replicating work unnecessarily, just not sure how to approach this and how to allocate the necessary resources.

**Comment by kellertuer** [comment]
Ok. I can not do much about other people using AI, but I would prefer to just talk and interact with people – it would be kind to respect that.

Timewise – no worries. I have a semester starting soon – and also many other things can affect response times. Sometimes topics also justo not fit and then they simply do not happen.

For concrete steps, I would propose start with basics, and maybe also just in one area of this package. I do not have a good overview on the ecosystem of packages, so maybe the following idea has side effects I do not see.

If you look at the manifold definition

https://github.com/JuliaGNI/GeometricOptimizers.jl/blob/5b5c96d5cf157374be511227e4180ddc4885e1fc/src/manifolds/abstract_manifold.jl#L1-L6

one could instead use a dependency on our (on purpose very lightweight) ManifoldsBase.jl and then use 

https://github.com/JuliaManifolds/ManifoldsBase.jl/blob/24e74d8e56cfdc67b5e8d9ffe9fbe8080a8a6f2d/src/maintypes.jl#L1-L20

instead, i.e. replace Stiefel and Grassmann from here also with their counterparts in Manifolds.jl: [Stiefel](https://juliamanifolds.github.io/Manifolds.jl/stable/manifolds/stiefel/) and [Grassmann](https://juliamanifolds.github.io/Manifolds.jl/stable/manifolds/grassmann/)

My problem there is, that for example in https://github.com/JuliaGNI/GeometricOptimizers.jl/blob/main/src/manifolds/stiefel_manifold.jl

* I do not know what is stored in the manifold (the field A is not documented)
* I am unsure what the `*` overloaded functions are
* I miss something like a retraction
* `rgrad` is basically https://juliamanifolds.github.io/Manifolds.jl/stable/manifolds/stiefel/#ManifoldsBase.change_representer-Tuple{Stiefel,%20EuclideanMetric,%20Any,%20Any} – though we split that a bit and it is recommended to use `riemmanian_gradient` instead (see https://juliamanifolds.github.io/ManifoldDiff.jl/stable/library/#ManifoldDiff.riemannian_gradient-Tuple{AbstractManifold,%20Any,%20Any})
* `metric(X,Y)` would change to `inner(M, p, X, Y)` which on Stiefel would just be the Euclidean metric from the embedding by default, but you can wrap the manifold in a `MetricManifold` with the canonical metric to get https://juliamanifolds.github.io/Manifolds.jl/stable/manifolds/stiefel/#ManifoldsBase.inner-Tuple{MetricManifold{ℝ,%20%3C:Stiefel{ℝ},%20CanonicalMetric},%20Any,%20Any,%20Any} which is what you documented to use. But you can also use the whole range in between then
* `global_section` we do not have since we never needed that before.
* the remaining code is a bit hard to comment on again, since it is undocumented – I could only issue guesses here.

I am probably missing quite a few points, but that is roughly where one could check for a start I think

Starting on the “other end” with the optimization is probably the wrong end.

**Comment by michakraus** [comment]
> Ok. I can not do much about other people using AI, but I would prefer to just talk and interact with people – it would be kind to respect that.

We totally understand and respect that. This was a mistake caused by a prompt along the lines of „mark outdated issues“ which the AI understood in a time-wise manner instead of „already fixed“. However, this was (a) caught in review, and (b) happened during an evaluation of a model (gpt-5.6-luna) we do not yet have much experience with.

> Timewise – no worries. I have a semester starting soon – and also many other things can affect response times. Sometimes topics also justo not fit and then they simply do not happen.

Thanks a lot for understanding!

> For concrete steps, I would propose start with basics, and maybe also just in one area of this package. I do not have a good overview on the ecosystem of packages, so maybe the following idea has side effects I do not see.

Thanks a lot for all the explanations. That helps us a lot to get going. We will try to put a bit more effort into this in the next few weeks. Benedikt will probably chime in, but I think it might be a good idea to start with moving our interfaces a bit closer to yours and improve on documentation. That should then make it easier to swap out structures from GeometricOptimizers with those from ManifoldsBase.

**Comment by kellertuer** [comment]
Sure, I do understand, mistakes happen :) 

The next 2-3 weeks I might be a bit busy (new semester with a new lecture ...) – but again there is overall no hurry.
If there are then a bunch of questions, we could also arrange for a digital meeting if that helps.
In general that could even be in German (your names both sound German – and that is my mother tongue) – unless we directly want to deep dive into GPU-things and would need Mateusz on board as well – then it would be in English.

**Comment by benedict-96** [comment]
Hallo @kellertuer, I am also really sorry for what happened; AI tools are greatly improving our efficiency when it comes to mechanically implementing code but it did lead to recklessness in that particular case.

You have pointed out some real issues with our code that we have to tackle sooner or later. Subtyping manifolds from `AbstractMatrix` can be seen as a historical artifact. It seemed to make sense at the time (we just needed a quick implementation of matrix manifolds for their use in neural networks). Subtyping from `AbstractManifold` instead should be the much smarter choice here!

Our aim with `GeometricOptimizers` up to now was to have a working implementation so that the optimizer functionality can be moved out from [`GeometricMachineLearning`](https://github.com/JuliaGNI/GeometricMachineLearning.jl), where it was up to now and which we use `GeometricOptimizers` for. Since this first step of having a working implementation is more or less done now, we will slowly venture into improving the package itself, or potentially replacing it altogether. What counts for us most is that the networks in `GeometricMachineLearning` can be trained efficiently on various GPU architectures.

A meeting would definitely make a lot of sense. @michakraus is still in a different time zone, which might make it somewhat complicated, but he will also be back in Germany in a few weeks!

**Comment by kellertuer** [comment]
> AI tools are greatly improving our efficiency when

Sorry, I have to disagree on this point. For me it has decreased efficiency because I too often have to argue and fight against other peoples slop. So if “everyone else” gets faster at the cost of a higher workload on my side, ... I thought about leaving FOSS a few times now, but decided, I like my package and will not do that – but I will not tolerate if people steal my time with AI slop.

I am not saying AI is useless – I have not used it much myself yet, mainly for ethical and personal reasons; It might maybe help unifying things, small refactors – but if it is at the cost of other peoples increase of work – that is not what it should do.
So if you prefer to vibe-code, Manopt.jl is not for you. Until now, typos and a but of unification of documentation was done there with a bit of help, but the rest I prefer to do – and understand the code – myself.

And sure there is no hurry – and if you notice it does not fit, either because it is too much work to do for GPUs on our side still or you get angry at me not liking all the AI here – then it maybe also does simply not fit, that is fine with me as well.
But we can surely wait for Michael to be back in a nicer timezone anyways, sure.

**Comment by benedict-96** [comment]
I totally understand your frustration, I've heard terrible stories from people sitting on the receiving end of AI slop-based pull requests!

We definitely look into ManifoldsBase.jl and Manopt.jl. 
In fact, long before AI had any practical use at all, we actually had `Manifolds.jl` as a dependency in `GeometricMachineLearning`: https://github.com/JuliaGNI/GeometricMachineLearning.jl/commit/b0dde8762e3ed2688d4d93c54ac34ab582ad669c.
In the process of developing the package though, we made a decision at some point to first have a working implementation of manifold-based neural networks that also run on GPUs at the expense of best FOSS practices and developed the lightweight optimizers that are now `GeometricOptimizers.jl`.
`GeometricOptimizers` is not something we are married to, if its functionality can be replicated somewhere else, we would be very happy to move it there.
